Formal2024/Workshops/workshop4.lean

```diff
@@ -0,0 +1,251 @@
+
+
+-- Apart from `∀` (which we discussed is equivalent to functions) everything else is constructed
+-- starting from here (mostly using Inductive Types, as we will see).
+end
+
+
+section
+/-
+  # How is a type different from a set?
+-/
+
+def setNat : Set ℕ := univ -- This is the set of all natural numbers
+
+-- Sets come with a predicate: `∈` (`\in` or `\mem`). Its negation is `∉` (`\notin`)
+
+#check ( 1 ∈ setNat ) -- This is in `Prop`
+#check ( 1 : ℕ )     -- This is not
+
+-- `( 1 : ℕ )` is a statement in type theory, but not an element in `Prop`, our model of
+-- propositional logic inside type theory.
+
+-- Another big difference: if `a : α` and `a : β`, then `α` and `β` are definitionally equal
+-- To make sense of things like subsets, intersections, union, complements, we need `Set`
+
+end
+
+section
+variable {α : Type*}
+variable (s t u : Set α)
+
+-- We write `⊆` (`\ss` or `\sub`) for subset, `∩` (`\i` or `\cap`) for intersection,
+-- `∪` (`\u` or `\cup`) for union, and `\` (typed as `\`) for set difference.
+
+-- We can also use the notation `{ y | P y }`, where `P` is a predicate, for the set of
+-- elements satisfying `P`
+
+#check subset_def
+#check mem_setOf
+#check inter_def
+#check union_def
+#check mem_diff
+
+example (h : s ⊆ t) : s ∩ u ⊆ t ∩ u := by
+  simp only [subset_def, inter_def, inter_def, mem_setOf]
+  rw [subset_def] at h
+  -- We can use `simp only` or `rw` to unfold the definitions
+  -- `simp only` is a more powerful `rw`, that also works under quantifiers
+  -- Now we can proceed as usual
+  rintro x ⟨xs, xu⟩
+  exact ⟨h x xs, xu⟩
+  tada
+
+example (h : s ⊆ t) : s ∩ u ⊆ t ∩ u := by
+  -- This is not needed though, as the anonymus constructor forces Lean to unfold the definitions
+  rintro x ⟨xs, xu⟩
+  exact ⟨h xs, xu⟩ -- There may be some slight differences (here we do not need to specify x)
+  tada
+
+
+example : s ∩ (t ∪ u) ⊆ s ∩ t ∪ s ∩ u := by
+  intro x hx                      -- Subset works like an implication
+  have xs : x ∈ s := hx.1         -- Intersection works like an `∧`
+  have xtu : x ∈ t ∪ u := hx.2    -- Union works like an `∨`
+  rcases xtu with xt | xu
+  · left
+    show x ∈ s ∩ t
+    exact ⟨xs, xt⟩
+  . right
+    show x ∈ s ∩ u
+    exact ⟨xs, xu⟩
+  tada
+
+example : (s \ t) \ u ⊆ s \ (t ∪ u) := by
+  intro x xstu
+  have xs : x ∈ s := xstu.1.1   -- `x ∈ s \ t` means `x ∈ s ∧ x ∉ t`
+  have xnt : x ∉ t := xstu.1.2
+  have xnu : x ∉ u := xstu.2
+  constructor
+  · exact xs
+  intro xtu
+  -- x ∈ t ∨ x ∈ u
+  rcases xtu with xt | xu
+  · show False; exact xnt xt
+  . show False; exact xnu xu
+  tada
+
+-- This is quite lengthy. We can use `rintro` to shorten the proof
+
+example : (s \ t) \ u ⊆ s \ (t ∪ u) := by
+  rintro x ⟨⟨xs, xnt⟩, xnu⟩
+  use xs
+  rintro (xt | xu) <;> contradiction -- We often need brackets when using `rintro` on a conjunction
+  tada
+
+-- Two sets are equal when they have the same elements. This is called "extensionality"
+-- You can use it by typing `ext _`
```

Comment:
We met the `ext` tactic in the last homework to show that two functions are the same. Maybe we should say more generally what `ext` is doing, i.e. apply lemmas marked with the `@[ext]` attribute, which are used to show to things are equal.

Reply:
I'll add some comment, and then maybe we can be more precise once we go into structures?

Reply:
Yeah that works!